### PR TITLE
MEN-5322: implement the IoT Manager Device State management APIs

### PR DESCRIPTION
**Pull request by tranchitella**
Changelog: title

Signed-off-by: Fabio Tranchitella <fabio.tranchitella@northern.tech>

### CODE REVIEW

File: api/http/management_state.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+		return
+	}
+
+	deviceID := c.Param(paramDeviceID)
+	if deviceID == "" {
+		rest.RenderError(c, http.StatusBadRequest, ErrEmptyDeviceID)
+		return
+	}
+	integrationID := c.Param(paramIntegrationID)
+	_ = integrationID
```
Comment (kjaskiewiczz) [inline api/http/management_state.go:82]: what does that do?

File: api/http/router.go [hunk 2]
```diff
@@ -47,10 +47,11 @@ const (
...
 
-	APIURLSettings      = "/settings"
-	APIURLDevice        = "/devices/:id"
-	APIURLDeviceTwin    = "/devices/:id/twin"
-	APIURLDeviceModules = "/devices/:id/modules"
+	APIURLSettings = "/settings"
+
+	APIURLDevice                 = "/devices/:id"
+	APIURLDeviceState            = APIURLDevice + "/state"
+	APIURLDeviceStateIntegration = APIURLDevice + "/state/:integrationId"
```
Comment (kjaskiewiczz) [inline api/http/router.go:54]: should integrationId be a path or query param?

File: app/app.go [hunk 3]
```diff
@@ -181,3 +190,46 @@ func (a *app) DeleteIOTHubDevice(ctx context.Context, deviceID string) error {
...
+func (a *app) GetDeviceStateIntegration(
+	ctx context.Context,
+	deviceID string,
+	integrationID string,
+) (*model.DeviceState, error) {
+	integration, err := a.store.GetIntegration(ctx, integrationID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve the integration")
+	} else if integration == nil {
+		return nil, ErrIntegrationNotFound
```
Comment (kjaskiewiczz) [inline app/app.go:207]: I think it's not handled by the management handlers

File: app/app.go [hunk 4]
```diff
@@ -181,3 +190,46 @@ func (a *app) DeleteIOTHubDevice(ctx context.Context, deviceID string) error {
...
+	}
+	switch integration.Provider {
+	case model.AzureIoTHub:
+		return a.GetDeviceStateIoTHub(ctx, deviceID, integration)
+	default:
+		return nil, ErrUnknownIntegration
```
Comment (kjaskiewiczz) [inline app/app.go:213]: I think it's not handled by the management handlers

File: app/app.go [hunk 5]
```diff
@@ -181,3 +190,46 @@ func (a *app) DeleteIOTHubDevice(ctx context.Context, deviceID string) error {
...
+	ctx context.Context,
+	deviceID string,
+	integrationID string,
+	state *model.DeviceState,
+) (*model.DeviceState, error) {
+	integration, err := a.store.GetIntegration(ctx, integrationID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve the integration")
+	} else if integration == nil {
+		return nil, ErrIntegrationNotFound
```
Comment (kjaskiewiczz) [inline app/app.go:227]: I think it's not handled by the management handlers

File: app/app.go [hunk 6]
```diff
@@ -181,3 +190,46 @@ func (a *app) DeleteIOTHubDevice(ctx context.Context, deviceID string) error {
...
+	}
+	switch integration.Provider {
+	case model.AzureIoTHub:
+		return a.SetDeviceStateIoTHub(ctx, deviceID, integration, state)
+	default:
+		return nil, ErrUnknownIntegration
```
Comment (kjaskiewiczz) [inline app/app.go:233]: I think it's not handled by the management handlers

File: model/device.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+// DeviceIntegration stores the integration details for a specific device
+type DeviceIntegration struct {
+	// ID is the integration ID
+	ID string `json:"id" bson:"id"`
+	// ConnectionString is the device's primary connection string
+	// nolint:lll
+	PrimaryConnectionString *ConnectionString `json:"primary_connection_string" bson:"primary_connection_string"`
+	// Secondary ConnectionString is the device's secondary connection string
+	// nolint:lll
+	SecondaryConnectionString *ConnectionString `json:"secondary_connection_string" bson:"secondary_connection_string"`
```
Comment (alfrunes) [inline model/device.go:34]: Do we need to store device credentials in this service?

File: model/device.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+	// ID is the device ID
+	ID string `json:"id" bson:"_id"`
+	// Integrations contains the list of integrations for this device
+	Integrations []DeviceIntegration `json:"integrations" bson:"integrations"`
+}
+
+// DeviceIntegration stores the integration details for a specific device
+type DeviceIntegration struct {
+	// ID is the integration ID
+	ID string `json:"id" bson:"id"`
```
Comment (alfrunes) [inline model/device.go:28]: #9 is using `uuid.UUID` for integration IDs.

File: model/device.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+//    limitations under the License.
+
+package model
+
+// Device represents a device and its available integrations
+type Device struct {
+	// ID is the device ID
+	ID string `json:"id" bson:"_id"`
+	// Integrations contains the list of integrations for this device
+	Integrations []DeviceIntegration `json:"integrations" bson:"integrations"`
```
Comment (alfrunes) [inline model/device.go:22]: We will probably at some point need to be able to query devices that belong to a given integration, in that case it we can do this efficiently if we have a multi key index on a field holding the list of integration IDs.
If we need additional metadata stored within the device document, I would propose adding another embedded document which maps integration IDs to meta-data.
```suggestion
	// Integrations contains the list of integrations for this device
	IntegrationIDs []uuid.UUID `json:"?" bson:"integration_ids"`
	
	Integrations map[string]MetaData `json:"integrations,omitempty" bson:"integrations,omitempty"`
```

File: model/device.go (hunk 7 above)
Comment (tranchitella) [reply model/device.go:34]: No, we don't, but I thought you want.
Let me remove it :)

File: model/device.go (hunk 9 above)
Comment (tranchitella) [reply model/device.go:22]: We don't need the meta data, let's go with the IntegrationIDs only.

File: api/http/management_state.go (hunk 1 above)
Comment (tranchitella) [reply api/http/management_state.go:82]: left-over, sorry

File: api/http/router.go (hunk 2 above)
Comment (tranchitella) [reply api/http/router.go:54]: According to @alfrunes's docs, it's a path param.

**Comment by tranchitella** [comment]
@alfrunes @kjaskiewiczz thanks for your reviews, ready for a second round.

I addressed all your concerns and requests. There is an additional change: now both GetDeviceStateIntegration and SetDeviceStateIntegration will look up the device by (device ID, integration ID) before proceeding with the calls to IoT Hub. This ensures that the device exists in Mender before issuing API requests to IoT hub with arbitrary device IDs.

**Review by kjaskiewiczz — APPROVED** [review]
looks good

File: store/mongo/datastore_mongo.go [hunk 10]
```diff
@@ -36,15 +37,21 @@ import (
...
 
 	ConnectTimeoutSeconds = 10
 	defaultAutomigrate    = false
 )
 
 var (
-	ErrFailedToGetSettings = errors.New("Failed to get settings")
+	ErrFailedToGetDevice      = errors.New("Failed to get device")
+	ErrFailedToGetIntegration = errors.New("Failed to get integration")
+	ErrFailedToGetSettings    = errors.New("Failed to get settings")
```
Comment (alfrunes) [inline store/mongo/datastore_mongo.go:54]: Nitpick: Error messages should not be capitalized.

**Review by alfrunes — APPROVED** [review]

File: store/mongo/datastore_mongo.go [hunk 11]
```diff
@@ -265,3 +269,38 @@ func (db *DataStoreMongo) CreateIntegration(
...
 	return err
 }
+
+func (db *DataStoreMongo) GetDevice(ctx context.Context, deviceID string) (*model.Device, error) {
+	return db.GetDeviceByIntegrationID(ctx, deviceID, uuid.Nil)
+}
+
+func (db *DataStoreMongo) GetDeviceByIntegrationID(
+	ctx context.Context,
+	deviceID string,
```
Comment (alfrunes) [inline store/mongo/datastore_mongo.go:279]: Unused argument

File: store/mongo/datastore_mongo.go [hunk 12]
```diff
@@ -265,3 +269,38 @@ func (db *DataStoreMongo) CreateIntegration(
 
 	return err
 }
+
+func (db *DataStoreMongo) GetDevice(ctx context.Context, deviceID string) (*model.Device, error) {
+	return db.GetDeviceByIntegrationID(ctx, deviceID, uuid.Nil)
```
Comment (alfrunes) [inline store/mongo/datastore_mongo.go:274]: This function is not doing what it is intended for because deviceID is unused.

**Review by alfrunes — COMMENTED** [review]
I'm currently rebasing my work on top of yours and I found a significant error in the database interface. I can fix those in my upcoming PR.

File: store/mongo/datastore_mongo.go (hunk 12 above)
Comment (tranchitella) [reply store/mongo/datastore_mongo.go:274]: right, sorry, I fixed it!

**Comment by tranchitella** [comment]
> I'm currently rebasing my work on top of yours and I found a significant error in the database interface. I can fix those in my upcoming PR.

I force-pushed, sorry for the trouble.

**Pull request merged**